rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,266 @@
+# Constant Tensor Cache Capacity Control in Graph API
+
+## Introduction
+Currently we have a constant tensor cache feature in DNNL backend, which is used
+to cache those processed constant tensors (like reordered constant weights,
+folded constant scales, ...) to reduce redundant computation and improve
+performance. The following figure shows a basic design of this existing constant
+tensor cache: we will cache the internal processed constant tensor to the cache
+in the first iteration, and reuse it in next iterations. In concurrent execution
+scenarios, the same compiled partition running in different threads will share
+the same cached tensor, and we had a promise-future mechanism to allow only one
+thread to initiate the computation of constant tensor, and other threads to wait
+until the computation completes.
+
+![DNNL backend existing constant tensor cache](images/dnnl-backend-constant-cache.png)
+
+We have exposed two APIs to enable/disable the constant tensor cache and return
+the enabling status:
+
+```cpp
+void set_constant_tensor_cache(int flag);
+int get_constant_tensor_cache();
+```
+
+The existing constant tensor cache feature is disabled by default which means
+zero capacity on both CPU and GPU. In some memory limited scenarios, users
+may only want to cache a part of constant tensors to improve performance, but
+don't want the cached capacity to grow infinitely and consume too much memory.
+In this document, we propose to introduce a constant tensor cache capacity
+control API to address above requests.
+
+## Related works
+
+### TensorRT
+
+Users can control the max workspace size used inside TensorRT through the
+`nvinfer1::IBuilderConfig::setMemoryPoolLimit` and
+`nvinfer1::IBuilderConfig::getMemoryPoolLimit` APIs. If the workspace size is
+not big enough, TensorRT will show warning message to suggest users to increase
+the workspace size. If OOM occurs, users could also reduce workspace memory
+consumption through this API. And users are suggested to set the workspace size
+as big as possible to achieve best performance. The default workspace size is
+set to device global memory size from v8.4 release.
+
+## Proposal
+
+### Dedicated cache for different engine kind
+In oneDNN Graph programming model, different graph can have different engine
+kind and simultaneously run on different engine, like cpu or gpu. Different kind
+of engine usually has different memory limitation. 
+
+So, we propose to use dedicated cache for different engine kind, and let users
+set different capacity for different engine kind separately.
+
+For multiple devices scenarios, like multi-GPU, whether to use dedicated cache
+for each device is still an open question, we will discuss it in later sections.
+
+### Unified cache for all backends
```

Comment:
What does this paragraph serve for?

Reply:
This paragraph explains that the new API in the proposal does not differentiate between backends when setting capacity, which means that for the same hardware, different backends share the entire capacity。